### ISSUE TITLE
Implement local space particles

### ISSUE THREAD

**Issue by fadookie** [fadookie/particleman#10]
Particles that move automatically when emitter moves

# McHorse said
In my mod, I just capture the MV [Model View] matrix after translation and inverse it at point of rendering to find the position and orientation relative to lastGlobal

First I read the model view matrix from OpenGL: https://github.com/mchorse/blockbuster/blob/1.12/src/main/java/mchorse/blockbuster/client/render/RenderCustomModel.java#L134

Then I calculate the relative position here: https://github.com/mchorse/blockbuster/blob/1.12/src/main/java/mchorse/blockbuster_pack/morphs/SnowstormMorph.java#L187
And once the rendering of the entity is done, I release it (because in my mods I have nested entity rendering, etc, fun stuff): https://github.com/mchorse/blockbuster/blob/1.12/src/main/java/mchorse/blockbuster/client/render/RenderCustomModel.java#L123

> Can't you just translate using entites MV matrix and render all particles relative to origin?
I don't think you can do that
The MV matrix is always relative to the camera
Not to the world

> If local space is off-off, then just the `lastGlobal` is used **initially**
Initially, as in it will be set once, and from there the future calculation will be done
If local space on-off, then the position within particles will be global, but also they will be rotated using the `emitter.rotation`, and the motion will be also applied from that rotation, but if local space on-on, the position will be relative to local 0,0,0, and the particles will be rendered relative to the last global and rotation always
Basically whatever I shown in the video, I'm already confused myself

# Eliot said
ugh local space particles are definitely jacked up. getting this renderer code set up is gonna be a project but at least I have a blueprint thanks to McHorse
so i think... users are gonna have to create custom renderers (or mixin to vanilla ones) in order to get this local space matrix translation thing working properly
but hopefully it can just be set up to forward a single function call and i can handle the rest in the rendering helper
